src/tests/dotnet-monitor/PipelineTests.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.Contracts;
+using Microsoft.Diagnostics.Tracing.Parsers.Kernel;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace DotnetMonitor.UnitTests
+{
+    public class PipelineTests
+    {
+        private readonly ITestOutputHelper _outputHelper;
+
+        public PipelineTests(ITestOutputHelper outputHelper)
+        {
+            _outputHelper = outputHelper;
+        }
+
+        [Fact]
+        public async Task TestStartStopCancelDispose()
+        {
+            var timePipeline = new DelayPipeline();
+            var cancellationTokenSource = new CancellationTokenSource();
+            var token = cancellationTokenSource.Token;
+
+            await Assert.ThrowsAsync<PipelineException>(() => timePipeline.StopAsync());
+
+            var startTask = timePipeline.RunAsync(token);
+            var secondStartCall = timePipeline.RunAsync(token);
+            Assert.Equal(startTask, secondStartCall);
+
+            var timeoutSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+            var stopTask = timePipeline.StopAsync(timeoutSource.Token);
+            var secondStopCall = timePipeline.StopAsync(timeoutSource.Token);
+            Assert.Equal(stopTask, secondStopCall);
+
+            await Assert.ThrowsAnyAsync<OperationCanceledException>(() => stopTask);
+
+            cancellationTokenSource.Cancel();
+            
+            await Assert.ThrowsAnyAsync<OperationCanceledException>(() => startTask);
+
+            await timePipeline.DisposeAsync();
+
+            await Assert.ThrowsAsync<ObjectDisposedException>(() => timePipeline.RunAsync(token));
+
+            Assert.Equal(1, timePipeline.ExecutedAbort);
+        }
+
+        [Fact]
+        public async Task TestStart()
+        {
+            var timePipeline = new DelayPipeline(TimeSpan.Zero);
+            await timePipeline.RunAsync(CancellationToken.None);
+            await timePipeline.DisposeAsync();
+
+            Assert.Equal(0, timePipeline.ExecutedAbort);
```

Comment:
To clarify the behavior:
- cancellations from Start/Stop both trigger an Abort callback.
- Calling Dispose also causes a cancellation.

The current Dispose call is actually very similar to the previous Abort call, except that all stopping behavior relies on Cancellation. The current Abort call is protected and cannot be accessed externally. We could remove it and rely on Dispose for any sort of cleanup as well.

Reply:
> We could remove it and rely on Dispose for any sort of cleanup as well.

Could we also do the reverse, eliminate IAsyncDisposable and the OnDispose() API? So far I haven't recognized any scenario where implementing IAsyncDisposable is helping the pipeline user accomplish a useful task. The rules for cleanup don't seem that different than any other Task:
  - user doesn't call RunAsync() so no work has been started. Releasing the ref to the pipeline will cause everything to be GC'ed
  - user calls RunAsync() and awaits the Task so all work is complete when await returns. Releasing the ref to the pipeline causes everything to be GC'ed
  - user calls RunAsync() and fails to await the Task so the work continues even if user doesn't observe it. Releasing the ref to the pipeline might leak it indefinitely, similar to many other Tasks. If we assume not awaiting the Task was a mistake, it seems about as probable of a mistake as forgetting to use 'using'.

Reply:
The scenario I wanted to enable was:
- User creates pipeline, calls run async.
- Something else happens that requires the user to simply destroy everything, but the cancellation token may not be triggered. The user can force this by calling DisposeAsync. This will trigger a cancellation in the existing Run/Stop tokens, a possible callback to OnAbort, followed by a guaranteed callback to OnDispose.
- I think deterministic cleanup for pipelines is desirable.

Reply:
Per discussion, we now have just one final state that Run/Stop/Dispose all transition to, with one callback to override.